File: 0455. Assign Cookies/0455.py
# time complexity: O(n log n + m log m) -> Sorting both arrays dominates
# space complexity: O(1) -> No extra space 
class Solution(object):
    def findContentChildren(self, g, s):
        """
        :type g: List[int]
        :type s: List[int]
        :rtype: int
        """
        # sort the arrays
        g.sort()
        s.sort()
        
        child = 0 # pointer for children
        cookie = 0 # pointer for cookies

        while child < len(g) and cookie < len(s):
            if s[cookie] >= g[child]: # if cookies satisfies the greed
                child += 1
                cookie +=1
            else:                     # if not, next cookie, same child
                cookie +=1
        
        return child # number of content children

File: 0455. Assign Cookies/0455.java
// time complexity: O(nlogn + mlogm) -> sorting
// space complexity: O(1)
class Solution {
    public int findContentChildren(int[] g, int[] s) {
        // g -> greed factors
        // s -> size of the cookies

        // first, sort the arrays
        Arrays.sort(g);
        Arrays.sort(s);

        int output = 0; // count of content children
        int i = 0; // pointer for children
        int j = 0; // pointer for cookies

        while (i < g.length && j < s.length) {
            if (g[i] <= s[j]) { // if cookie satisfies greed
                i++; // move to next child
                output++; // +1 content children count
            }
            j++; // next cookie
        }

        return output;
    }
}